### PR TITLE
Add `TRANSIENT_ATTACHMENT` GPUTextureUsage

### PR DESCRIPTION
**Pull request by beaufortfrancois**
Following https://github.com/gpuweb/gpuweb/pull/5446, this PR adds `TRANSIENT_ATTACHMENT` GPUTextureUsage to the WebGPU API spec.

### CODE REVIEW

**Review by mwyrzykowski — APPROVED** [review]

File: spec/index.bs [hunk 1]
```diff
@@ -12497,35 +12185,44 @@ dictionary GPURenderPassDepthStencilAttachment {
...
+    1. Let |usage| be |this|.{{GPURenderPassDepthStencilAttachment/view}}.{{GPUTextureView/[[descriptor]]}}.{{GPUTextureViewDescriptor/usage}}.
+    1. All of the requirements in the following steps |must| be met.
+        <div class=validusage>
+            1. |this|.{{GPURenderPassDepthStencilAttachment/view}} |must| have a [=depth-or-stencil format=].
+            1. |this|.{{GPURenderPassDepthStencilAttachment/view}} |must| be a [$renderable texture view$].
+            1. If |this|.{{GPURenderPassDepthStencilAttachment/depthLoadOp}} is {{GPULoadOp/"clear"}},
+                |this|.{{GPURenderPassDepthStencilAttachment/depthClearValue}} |must| [=map/exist|be provided=] and |must| be between 0.0 and 1.0,
+                inclusive.
+                <!-- POSTV1(unrestricted-depth): unless unrestricted depth is enabled -->
+            1. If |format| has a depth aspect and |this|.{{GPURenderPassDepthStencilAttachment/depthReadOnly}} is `false`:
```
Comment (kainino0x) [inline spec/index.bs]: It should also be invalid to use `depthReadOnly`/`stencilReadOnly` (which, behaviorally, implies `"load"`/`"store"`) with a texture with `TRANSIENT_ATTACHMENT`

Comment (mwyrzykowski) [reply spec/index.bs]: I agree it should be invalid, but I think that's already covered by (currently line 12206):
> If |usage| includes the {{GPUTextureUsage/TRANSIENT_ATTACHMENT}} bit:
                    1. |this|.{{GPURenderPassDepthStencilAttachment/depthLoadOp}} |must| be {{GPULoadOp/"clear"}}.

because https://www.w3.org/TR/webgpu/#abstract-opdef-gpurenderpassdepthstencilattachment-gpurenderpassdepthstencilattachment-valid-usage says if `depthReadOnly` is true then `depthLoadOp must not be provided`

Wouldn't hurt to explicitly call it out

Comment (kainino0x) [reply spec/index.bs]: That's nested under "if `format` has a depth aspect and `depthReadOnly` is false", so it won't be checked if `depthReadOnly` is true.

Comment (mwyrzykowski) [reply spec/index.bs]: > If format has a depth aspect and this.[depthReadOnly](https://www.w3.org/TR/webgpu/#dom-gpurenderpassdepthstencilattachment-depthreadonly) is false:
> this.[depthLoadOp](https://www.w3.org/TR/webgpu/#dom-gpurenderpassdepthstencilattachment-depthloadop) must [be provided](https://infra.spec.whatwg.org/#map-exists).
this.[depthStoreOp](https://www.w3.org/TR/webgpu/#dom-gpurenderpassdepthstencilattachment-depthstoreop) must [be provided](https://infra.spec.whatwg.org/#map-exists).
>
> Otherwise:
> this.[depthLoadOp](https://www.w3.org/TR/webgpu/#dom-gpurenderpassdepthstencilattachment-depthloadop) must not [be provided](https://infra.spec.whatwg.org/#map-exists).
this.[depthStoreOp](https://www.w3.org/TR/webgpu/#dom-gpurenderpassdepthstencilattachment-depthstoreop) must not [be provided](https://infra.spec.whatwg.org/#map-exists).

~~It's under the `Otherwise` clause, i.e., format has no depth aspect or depthReadOnly is true~~

Comment (mwyrzykowski) [reply spec/index.bs]: Sorry this whole block is under the "if format has a depth aspect and depthReadOnly is false", my apologies.

Yes we need to call it out then in the below section

Comment (mwyrzykowski) [reply spec/index.bs]: Or alternatively we move the:

> If |usage| includes the {{GPUTextureUsage/TRANSIENT_ATTACHMENT}} bit:
                    1. |this|.{{GPURenderPassDepthStencilAttachment/depthLoadOp}} |must| be {{GPULoadOp/"clear"}}.
                    1. |this|.{{GPURenderPassDepthStencilAttachment/depthStoreOp}} |must| be {{GPUStoreOp/"discard"}}.

restriction outside the `If |format| has a depth aspect and |this|.{{GPURenderPassDepthStencilAttachment/depthReadOnly}} is `false`:`

Comment (beaufortfrancois) [reply spec/index.bs]: I've followed @mwyrzykowski's suggestion. Let me know if that looks good to you.

Comment (Kangz) [reply spec/index.bs]: I'm concerned that this might prevent using stencil-only or depth-only transient attachments because IIRC the load/StoreOps of the aspect that's not present must be undefined. Maybe we should have remove only the check for "depth/stencilReadonly" in that PR but not the missing aspect check.

Comment (beaufortfrancois) [reply spec/index.bs]: Good catch! I've updated this PR with your feedback.

Comment (kainino0x) [reply spec/index.bs]: Thanks for the iteration. LGTM now

**Review by kainino0x — COMMENTED** [review]
Editorial LGTM. Could you also update the "Roadmap" in the new proposal document, as part of this PR?

**Comment by beaufortfrancois** [comment]
> Editorial LGTM. Could you also update the "Roadmap" in the new proposal document, as part of this PR?

Done.

**Comment by Kangz** [comment]
@teoxoy @mwyrzykowski PTAL (again)! Maybe we can get this approved offline without having to wait for the next meeting. (we'd merge only once the CTS is implemented and tests pass in a browser)

**Comment by mwyrzykowski** [comment]
> @teoxoy @mwyrzykowski PTAL (again)! Maybe we can get this approved offline without having to wait for the next meeting. (we'd merge only once the CTS is implemented and tests pass in a browser)

✅ from Apple's side

File: proposals/transient-attachments.md [hunk 2]
```diff
@@ -25,10 +24,15 @@ partial namespace GPUTextureUsage {
...
 
+The `GPUCanvasContext/configure(configuration)` algorithm is extended with the following change:
+
+- If `configuration.usage` includes the `TRANSIENT_ATTACHMENT` bit, throw a TypeError.
+
 The `validating GPUTextureDescriptor(this, descriptor)` algorithm is extended with the following change:
 
 - If `descriptor.usage` includes the `TRANSIENT_ATTACHMENT` bit:
   - `descriptor.usage` must contain only and exactly `TRANSIENT_ATTACHMENT` and `RENDER_ATTACHMENT` bits.
+  - `descriptor.dimension` must be `"2d"`.
```
Comment (kainino0x) [inline proposals/transient-attachments.md:33]: I would like to also require depthOrArrayLayers=1 and mipLevelCount=1 as I don't think they are useful and only make the implementation more fragile.

Comment (beaufortfrancois) [reply proposals/transient-attachments.md:33]: Done in https://github.com/gpuweb/gpuweb/pull/5450/commits/70c743b1958299b661386aadb13c566a300e827c

Comment (mwyrzykowski) [reply proposals/transient-attachments.md:33]: Is there a driver / hardware requirement for disallowing mips @kainino0x @beaufortfrancois ? 

It could be useful if you want to build a mip chain and have a corresponding transient MSAA texture. You can manually construct the transient mip chain but that seems unnecessary.

**Review by kainino0x — APPROVED** [review]

File: proposals/transient-attachments.md (hunk 2 above)
Comment (kainino0x) [reply proposals/transient-attachments.md:33]: I didn't have a specific requirement, just a feeling that it requires significant extra testing, for what is probably a little-used corner of native APIs that would be relatively likely to have driver bugs. Of course it would probably be workaround-able, but it didn't seem worth the effort. It's a fair point though that it's convenient so maybe the WG would prefer to go that direction. I'm OK either way with a preference for having more limitations at least to start.

Comment (mwyrzykowski) [reply proposals/transient-attachments.md:33]: Sounds good - it is fine with me to keep the restriction and then remove it if someone asks for it to be removed :)

**Comment by beaufortfrancois** [comment]
I believe we're now good for a final approval of this feature in WG meeting as we now have [CTS tests](https://github.com/gpuweb/cts/issues/4509) that pass with Chrome behind an experimental flag ([crbug](https://issues.chromium.org/issues/462620664))

**Review by penendangkatmukanak82-cell — APPROVED** [review]
